### PR TITLE
require shares provided to placeTrade & tradeUntilAmountIsZero

### PR DESCRIPTION
**Pull request by nuevoalex**
This is so someone doesn't have to front the full ETH value of their total orders when they're providing shares to satisfy some portion of the order.

The idea here being we get the `sharesDepleted` or `otherSharesDepleted` (whichever has !== '0' value) from the simulatedTrade and provide that to the `placeTrade` call as the shares provided.

On each iteration of tradeUntilAmountIsZero it will recalculate how many if any shares are still being provided.

Tested this in the UI with a balance too low to front the full cost and it seemed to work correctly but please make sure I'm not being naive with this.

### CODE REVIEW

File: src/trading/trade-until-amount-is-zero.js [hunk 1]
```diff
@@ -72,8 +75,12 @@ function tradeUntilAmountIsZero(p) {
           if (p.doNotCreateOrders) return p.onSuccess(tradeOnChainAmountRemaining.toFixed());
           return p.onFailed(new Error("Trade completed but amount of trade unchanged"));
         }
+        var newAmount = convertOnChainAmountToDisplayAmount(tradeOnChainAmountRemaining, tickSize);
```
Comment (tinybike) [inline src/trading/trade-until-amount-is-zero.js]: I think we should add `"sharesProvided"` to the `immutableDelete` array in `tradePayload` definition (https://github.com/AugurProject/augur.js/pull/391/files#diff-ca59c39682aec540f1296b9a53f3d384L55), so an argument of this name isn't passed to the underlying API.

**Comment by tinybike** [comment]
I'm getting the following error when I try to run the create-orders.sh script (with USE_PUBLIC_CREATE_ORDER=false) on a private chain:
```
/home/jack/src/augur.js/node_modules/bignumber.js/bignumber.js:1309
                    throw Error
                    ^

Error: [BigNumber Error] Not a number: undefined
    at /home/jack/src/augur.js/node_modules/bignumber.js/bignumber.js:1309:27
    at new BigNumber (/home/jack/src/augur.js/node_modules/bignumber.js/bignumber.js:220:67)
    at new BigNumber (/home/jack/src/augur.js/node_modules/bignumber.js/bignumber.js:233:25)
    at /home/jack/src/augur.js/src/trading/trade-until-amount-is-zero.js:79:88
    at /home/jack/src/augur.js/src/trading/get-trade-amount-remaining.js:44:5
    at parseEthereumResponse (/home/jack/src/augur.js/node_modules/ethrpc/src/decode-response/parse-ethereum-response.js:9:3)
    at responseHandler (/home/jack/src/augur.js/node_modules/ethrpc/src/rpc/blockchain-message-handler.js:36:7)
    at /home/jack/src/augur.js/node_modules/ethrpc/src/rpc/blockchain-message-handler.js:52:7
    at store.dispatch (/home/jack/src/augur.js/node_modules/redux-thunk-subscribe/src/index.js:10:18)
    at Object.<anonymous> (/home/jack/src/augur.js/node_modules/ethrpc/src/reset-state.js:45:9)
```

**Comment by tinybike** [comment]
Oh I see, `sharesProvided` is now a required parameter, that's why it's failing, let me fix the script.

File: src/trading/trade-until-amount-is-zero.js [hunk 2]
```diff
@@ -52,7 +55,7 @@ function tradeUntilAmountIsZero(p) {
     console.info("tradeUntilAmountIsZero complete: only dust remaining");
     return p.onSuccess(null);
   }
-  var tradePayload = assign({}, immutableDelete(p, ["doNotCreateOrders", "numTicks", "minPrice", "maxPrice"]), {
+  var tradePayload = assign({}, immutableDelete(p, ["doNotCreateOrders", "numTicks", "minPrice", "maxPrice", "sharesProvided"]), {
     tx: assign({ value: convertBigNumberToHexString(cost), gas: speedomatic.prefixHex(calculateTradeGas().toString(16)) }, p.tx),
```
[the remark this answers is not in the document]
Comment (justinbarry) [reply src/trading/trade-until-amount-is-zero.js]: sorry, should it be documented?

[the remark this answers is not in the document]
Comment (tinybike) [reply src/trading/trade-until-amount-is-zero.js]: It's a common input for all wrappers for API functions and the API functions themselves.  I'm not opposed to adding a jsdoc comment for it but we'd need to add it to basically every public function in augur.js for consistency.

[the remark this answers is not in the document]
Comment (justinbarry) [reply src/trading/trade-until-amount-is-zero.js]: Okay, good to know.

[the remark this answers is not in the document]
Comment (justinbarry) [reply src/trading/trade-until-amount-is-zero.js]: Okay, good to know.

**Comment by tinybike** [comment]
Hm, filling orders seems to work ok thru the UI, just the fill-order script not working.

**Comment by tinybike** [comment]
Ah ok @nuevoalex @JohnDanz the reason the script is failing is actually that the type of market is different.  Seems that create _and_ fill order are now failing on 8-outcome categoricals (haven't tested other numbers) with the following error:
```
wrap-on-failed-callback.js?cef0:11 transaction failed: {callReturn: "0x0000000000000000000000000000000000000000000000000000000000000000", blockHash: "0x274a0effa0be0efeeb64324016aa29935b0cd86fc738c26d55cc09b0cf23abcd", blockNumber: "0x9bd", from: "0xa7d1c6947d5a372747bca5fb449e52a22fd9ff28", gas: "0x5e3918", …} {constant: false, name: "publicTrade", inputs: Array(8), signature: Array(8), returns: "bytes32", …} 
```
Specific market I'm testing is the "Which city will have the highest median single-family home price in 2018?" canned market on a private chain.

**Comment by tinybike** [comment]
Hm, create market transactions are now failing too.  Now I'm not sure if maybe my local setup isn't borked somehow.  Nuking and retrying.

**Comment by tinybike** [comment]
@nuevoalex where does this value for minimum gas needed (on-contract) come from? https://github.com/AugurProject/augur-core/blob/master/source/contracts/trading/Trade.sol#L18

Is it possible we're just underestimating that value, and running out of gas...?

**Pull request merged**